### ISSUE TITLE
Nothing in .txt file when using darknet detector test

### ISSUE THREAD

**Issue by Victorsoeby** [AlexeyAB/darknet#8613]
I am using yolov4 with docker and have been trying to make this work using custom data. Currently i have been able to train properly, using command: 

```
 darknet detector train directory/to/obj.data directory/to/yolov4-tiny-one-class-training.cfg directory/to/yolov4-tiny.conv.29 -map 0.02 -dont_show directory/to/train_log.txt
```
This gives me loads of images output in my container with filenames aug_...jpg, with nice bounding boxes. The chart also shows, and shows the following: 

![chart](https://user-images.githubusercontent.com/56953590/181234884-f6f991ed-9cf2-41f2-a511-5464a27e72ac.png)

All in all, this looks fine, but when i then want to make predictions on my test set and see these results i struggle. I use the command: 

```
darknet detector test /directory/to/obj.data directory/to/yolov4-tiny-one-class-training.cfg directory/to/yolov4-tiny.conv.29 -ext_output -show_imgs -out results.txt < directory/to/test.txt
```
This does create the results.txt file, but it is empty. The terminal when running the command looks like this: 
![image](https://user-images.githubusercontent.com/56953590/181235755-671c56b9-0591-4690-a0a9-7c8d72fcf05f.png)

I think i've set it up correctly, as the test.txt file contains image_paths to my test set, just like my train and validation sets which are in the obj.data file.
I've also tried just making a prediction of one image (which should save as prediction.jpg), but this file isn't created either. Using something along the lines of this: 

```
darknet detector test /directory/to/obj.data directory/to/yolov4-tiny-one-class-training.cfg directory/to/yolov4-tiny.conv.29 directory/to/test_img.jpg
```
Hope someone can help me out.

**Comment by CemEntok** [comment]
Could there be a bug or something we are missing with using command? since I am also having errors when I used this:

> To process a list of images data/train.txt and save results of detection to result.txt use: ./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -dont_show -ext_output < data/train.txt > result.txt

My command is:
`$ cd /home/sureuser/objdetection/darknet`
`$ ./darknet detector test /home/sureuser/objdetection/datasetYolo/250722/obj.data /home/sureuser/objdetection/datasetYolo/250722/yolov4-obj.cfg /home/sureuser/objdetection/datasetYolo/250722/backup/yolov4-obj_best.weights -ext_output -dont_show -out result.txt < /home/sureuser/objdetection/datasetYolo/250722/test.txt`

And folders are here:
` $ cd /home/sureuser/objdetection/datasetYolo/250722`
`  $ cd ls`
Result is:
`backup  data_knife  obj.data  obj.names  result.txt  test.txt  train.txt  yolov4.conv.137  yolov4-obj.cfg`

Error is:
`can't open/read file: check file path/integrityeuser/objdetection/opencv_v0/opencv/modules/imgcodecs/src/loadsave.cpp (239) findDecoder imread_('data_knife/test/9_.jpg
Cannot load image data_knife/test/9_.jpg
 Detection layer: 139 - type = 28 
 Detection layer: 150 - type = 28 
 Detection layer: 161 - type = 28 
: Predicted in 9.548000 milli-seconds.`

test.txt is like that
`data_knife/test/1b5a6814bb670148.jpg`
`data_knife/test/1c60c3a3bbddc658.jpg`
`data_knife/test/0_.jpg`
`data_knife/test/100_.jpg`
`data_knife/test/101_.jpg`
`data_knife/test/102_.jpg`
`data_knife/test/103_.jpg`
`data_knife/test/104_.jpg`
...` and goes on.

A help would be appreciated.

**Comment by stephanecharette** [comment]
> I am using yolov4 with docker and have been trying to make this work using custom data. Currently i have been able to train properly, using command:
> 
> ```
>  darknet detector train directory/to/obj.data directory/to/yolov4-tiny-one-class-training.cfg directory/to/yolov4-tiny.conv.29 -map 0.02 -dont_show directory/to/train_log.txt
> ```

What does `0.02` do in that command?  The `-map` field doesn't take any parameters as far as I know.

Same thing with the `directory/to/train_log.txt` -- what does it do?  The `-dont_show` field also doesn't take any parameters.  Are you certain you're using the right commands?  Where did you get these commands from?

**Comment by stephanecharette** [comment]
> A help would be appreciated @AlexeyAB @stephanecharette

Why are you adding a new problem on top of a new ticket another user created?

Check your paths.  You have relative paths (versus absolute paths) which don't make sense due to the directory you are in.

**Comment by CemEntok** [comment]
Since it was almost the same kind of problem, i wrote but would it be more appropriate to open a new ticket about it?

Since darknet is built in `~/objdetection/darknet directory`;

 I am running that command in this directory but my files are in:

`/home/sureuser/objdetection/datasetYolo/250722` as you described [here](https://www.ccoderun.ca/programming/darknet_faq/#directory_setup)

So absolute paths should be also included in `test.txt` if I understood correctly. Now, I have tried this command

`./darknet detector test /home/sureuser/objdetection/datasetYolo/250722/obj.data /home/sureuser/objdetection/datasetYolo/250722/yolov4-obj.cfg /home/sureuser/objdetection/datasetYolo/250722/backup/yolov4-obj_best.weights -ext_output -dont_show < /home/sureuser/objdetection/datasetYolo/250722/test.txt > result.txt`

Before using this command, I have changed `test.txt` file which the paths of images are now **absolute** paths:

`/home/sureuser/objdetection/darknet/data_knife/test/1b5a6814bb670148.jpg`
`/home/sureuser/objdetection/darknet/data_knife/test/1c60c3a3bbddc658.jpg`
`/home/sureuser/objdetection/darknet/data_knife/test/0_.jpg`
`/home/sureuser/objdetection/darknet/data_knife/test/100_.jpg`
`/home/sureuser/objdetection/darknet/data_knife/test/101_.jpg`
`/home/sureuser/objdetection/darknet/data_knife/test/102_.jpg`
`/home/sureuser/objdetection/darknet/data_knife/test/103_.jpg`
`/home/sureuser/objdetection/darknet/data_knife/test/104_.jpg`
...

I am still having error

`can't open/read file: check file path/integrityopencv_v0/opencv/modules/imgcodecs/src/loadsave.cpp (239) findDecoder imread_('/home/sureuser/objdetection/darknet/data_knife/test/0_.jpg
Cannot load image /home/sureuser/objdetection/darknet/data_knife/test/0_.jpg`

I can open another ticket if it is more convenient way to ask question, let me know. Thanks for fast response.

**Comment by stephanecharette** [comment]
I'm not on your computer.  I cannot see your files.  But you need to figure out why the file `/home/sureuser/objdetection/darknet/data_knife/test/0_.jpg` cannot be opened.  Is the path correct?  Did you make a typo?  Is the file a valid image file?  Are there user permission issues?

Did you try simple things like `file /home/sureuser/objdetection/darknet/data_knife/test/0_.jpg` or `ls -l /home/sureuser/objdetection/darknet/data_knife/test/0_.jpg`?

Did you build a custom version of OpenCV, or did you install it as described here: https://www.ccoderun.ca/programming/darknet_faq/#opencv_and_cuda

And yes, you should have opened a different issue since it is an unrelated problem.

**Comment by Victorsoeby** [comment]
Hey, thanks for the quick reply. 
I think the 0.2 for the map flag alters the plot created durng training, i'll test that. The train_log.txt from my train command is a leftover from when i thought the training process created a logfile - sorry about that. My training command is working as intended, its more the darknet detector test command that is causing me issues, and not making any predictions. I recopied the command below, since this thread got a bit confusing, haha.
 
```
darknet detector test /directory/to/obj.data directory/to/yolov4-tiny-one-class-training.cfg directory/to/yolov4-tiny.conv.29 -ext_output -show_imgs -out results.txt < directory/to/test.txt
```
As stated earlier it doesn't seem to make any predictions, as the results.txt file is empty

**Comment by stephanecharette** [comment]
> I recopied the command below, since this thread got a bit confusing, haha.
> 
> ```
> darknet detector test /directory/to/obj.data directory/to/yolov4-tiny-one-class-training.cfg directory/to/yolov4-tiny.conv.29 -ext_output -show_imgs -out results.txt < directory/to/test.txt
> ```

Because you are using the command correctly.  Please re-read the example commands in the readme.  There should be zero references to yolov4-tiny.conv.29 in that command.

**Comment by Victorsoeby** [comment]
Oh god you're probably right. I will try using the yolov4-obj_best.weights file generated from the training when i get back on my pc tomorrow!

**Comment by Victorsoeby** [comment]
Hey again @stephanecharette, thanks for helping out this far, unfortunately it's still causing issues. 
After changing the weight file to the one outputted from training, and double checking some other examples, I am still unable to make any sorts of predictions. Test command is then: 
```
/abs/path/to/obj.data abs/path/to/yolov4-tiny-one-class-training.cfg abs/path/to/yolov4-tiny-one-class-training_best.weights -ext_output -dont_show -out results.txt < /abs/path/to/test.txt
```
As far as i can tell i am then replicating exactly what is in the readme, where it states: 
```
To process a list of images data/train.txt and save results of detection to result.txt use: darknet.exe detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -dont_show -ext_output < data/train.txt > result.txt
```
I am beginning to wonder if Docker is what is causing me the issue. I am mounting my data and pretrained weights into the container, which has darknet inside of it. This means that my file structure doesn't follow this example: https://towardsdatascience.com/custom-object-detection-using-darknet-9779170faca2 where they put their actual custom data within the darknet/data folder, weights in darknet/weights etc.. However i do believe it should be possible to just point to the correct directories, as this is working for my training procedure. Any other ideas?

**Comment by CemEntok** [comment]
> darknet detector test /abs/path/to/obj.data /abs/path/to/yolov4-tiny-one-class-training.cfg /abs/path/to/yolov4-tiny-one-class-training_best.weights -ext_output -dont_show -out results.txt < /abs/path/to/test.txt

Why you are using `-out` in the command? I didnt see this command in readme. Can you run the command discarding `-out` there? 
And what is your error message saying? Any screenshots?

**Comment by Victorsoeby** [comment]
I found the -out in another online guide, but removing it seems to do nothing either.. My terminal is outputting the following, where i've blurred out some of the paths: 
![image](https://user-images.githubusercontent.com/56953590/182334202-9ca04df5-b1da-4585-af36-1409b4d774cb.png)
The command thus reads: 
```
darknet detector test obj.data yolov4-tiny-one-class-training.cfg yolov4-tiny-one-class-training_best.weights -dont_show -ext_output < test_txt_file  > results.txt
```
Still no results.txt file.

**Comment by CemEntok** [comment]
I think the command should be like this

`...-dont_show -ext_output < test.txt > result.txt` 

However, you had `test_txt_file` without an extension, isn't it giving any errors?

**Comment by Victorsoeby** [comment]
> `...-dont_show -ext_output < test.txt > result.txt`
This gives the same situation as before, except the empty file is not created this time around. The test_file_txt is a variable which is defined elsewhere in the code as an absolute path. 
Still no errors present in the terminal.
